### PR TITLE
hashicorp commands

### PR DESCRIPTION
**Pull request by Shadowssong**

### CODE REVIEW

**Comment by fromonesrc** [comment]
I like where this is headed

**Comment by Shadowssong** [comment]
@articulate/web-ops Ready for :eyes: 

Tested each command locally and :+1:

File: scripts/hashicorp.coffee [hunk 1]
```diff
@@ -0,0 +1,143 @@
+# Description:
+#   Run any hashicorp job in jenkins 
+#
+# Commands:
+#   hubot ami create <type> <environment> <project> - Create ami type for a given project and environment
+#   hubot ami create <type> <environment> <project> from <github_ref> - Create ami type for a given project at a specific git ref
+#   hubot ami create <type> <environment> <project> role <chef_role> - Create ami but override role to specific role
+#   hubot ami deploy <project> to <environment>  - Run terraform apply and autoscale on a given environment and project
+#   hubot asg command <environment> <project> <command> - Run a command on a single node in the specified autoscaling group
+#   hubot asg autoscale <environment> <project> - Scale up and scale down a specified autoscaling group (to get new AMI's deployed)
```
Comment (fromonesrc) [inline scripts/hashicorp.coffee]: I worry there's room for error with this command. It allows a "deploy" to happen outside the context of a deploy. If someone wants a new version out, I'd like to see them running the general deploy workflow. Unless there's a strong reason to keep this chat-enabled I'd say we remove this option from botzo.

Comment (Shadowssong) [inline scripts/hashicorp.coffee]: Will remove :+1:

**Comment by tecnobrat** [comment]
We REALLY need to DRY these job thingies up.

**Comment by Shadowssong** [comment]
@tecnobrat I'm open to any advice, I tried to use regex and stuff to simplify it but it doesn't work considering different text would result is a different parameter being needed in the jenkins job.

File: scripts/hashicorp.coffee [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# Description:
+#   Run any hashicorp job in jenkins 
+#
+# Commands:
+#   hubot ami create <type> <environment> <project> - Create ami type for a given project and environment
+#   hubot ami create <type> <environment> <project> from <github_ref> - Create ami type for a given project at a specific git ref
+#   hubot ami create <type> <environment> <project> role <chef_role> - Create ami but override role to specific role
+#   hubot ami deploy <project> to <environment>  - Run terraform apply and autoscale on a given environment and project
+#   hubot asg command <environment> <project> <command> - Run a command on a single node in the specified autoscaling group
```
Comment (fromonesrc) [inline scripts/hashicorp.coffee]: What's the use case for this? Makes me a little worried but might be irrational.

Comment (Shadowssong) [inline scripts/hashicorp.coffee]: Any time someone needs to run a rake task or something like that

**Comment by tecnobrat** [comment]
@Shadowssong in general, not in this PR I don't think.

I'm thinking we implement some helper methods, like this:

```
robot.http("#{JENKINS_URL}/job/ami-build/buildWithParameters")
    .headers(Authorization: auth)
    .query(DEPLOY_TYPE: type, ENVIRONMENT: environment, PROJECT: project)
    .header("Content-Length", 0)
    .post() (err, res, body) ->
      if err
        msg.reply "Jenkins error: #{err}"
      else if 200 <= res.statusCode < 400
        msg.reply "Creating #{type} ami for #{environment}-#{project}"
      else
        msg.reply "Invalid response code: #{res.statusCode} #{body}"
```

Could be a function that is something like pseudocode:

```
createJenkinsJob: (job_name, auth, query_params, success, fail) ->
  robot.http("#{JENKINS_URL}/job/ami-build/buildWithParameters")
    .headers(Authorization: auth)
    .query(query_params)
    .header("Content-Length", 0)
    .post() (err, res, body) ->
      if 200 <= res.statusCode < 400
        success(res)
      else
        fail(res)
```

**Pull request merged**